src/cast_core/src/StreamStatsReporter.cpp

```diff
@@ -0,0 +1,115 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cast_core/actors/StreamStatsReporter.hpp>
+
+#include <memory>
+
+#include <yaml-cpp/yaml.h>
+
+#include <bsoncxx/array/view.hpp>
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/json.hpp>
+
+#include <metrics/operation.hpp>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/collection.hpp>
+#include <mongocxx/database.hpp>
+
+#include <boost/log/trivial.hpp>
+#include <boost/throw_exception.hpp>
+
+#include <gennylib/Cast.hpp>
+#include <gennylib/MongoException.hpp>
+#include <gennylib/context.hpp>
+
+#include <value_generators/DocumentGenerator.hpp>
+
+namespace genny::actor {
+
+namespace {
+
+// Poll interval for `streams_getStats`.
+static constexpr std::chrono::milliseconds kPollInterval{250ms};
+
+}; // namespace
+
+struct StreamStatsReporter::PhaseConfig {
+    mongocxx::database database;
+    bsoncxx::document::value statsCommand = bsoncxx::from_json("{}");
+
+    int64_t expectedDocumentCount{0};
+
+    PhaseConfig(PhaseContext& phaseContext, mongocxx::database&& db, ActorId id)
+        : database{db}, expectedDocumentCount{phaseContext["ExpectedDocumentCount"].to<int64_t>()} {
+        int threads = phaseContext.actor()["Threads"].to<int>();
+        if (threads != 1) {
+            BOOST_THROW_EXCEPTION(InvalidConfigurationException("StreamStatsReporter only allows 1 thread"));
+        }
+
+        auto repeat = phaseContext["Repeat"].maybe<int>();
+        if (!repeat || *repeat != 1) {
+            BOOST_THROW_EXCEPTION(InvalidConfigurationException("StreamStatsReporter only allows a single repeat"));
+        }
+
+        std::string streamProcessorName = phaseContext["StreamProcessorName"].to<std::string>();
+        bsoncxx::builder::stream::document builder;
+        statsCommand = builder << "streams_getStats" << "" << "name" << streamProcessorName << bsoncxx::builder::stream::finalize;
+    }
+};
+
+void StreamStatsReporter::run() {
+    for (auto&& config : _loop) {
+        for (const auto&& _ : config) {
+            auto throughput = _throughput.start();
+
+            int64_t inputMessageCount{0};
+            double inputMessageBytes{0};
+            while (true) {
+                auto reply = config->database.run_command(config->statsCommand.view());
+                BOOST_LOG_TRIVIAL(debug) << "stats reply: " << bsoncxx::to_json(reply);
+
+                inputMessageCount = reply.view()["inputMessageCount"].get_int64();
+                inputMessageBytes = reply.view()["inputMessageSize"].get_double();
+                if (inputMessageCount >= config->expectedDocumentCount) {
+                    break;
+                }
+
+                std::this_thread::sleep_for(kPollInterval);
```

Comment:
If we use the built-in SleepBefore/SleepAfter/GlobalRateLimit then we can get this for free and continue to participate in interrupt handling and the like.

Reply:
ahh but those built-ins currently require either `Repeat` or `Duration` to be defined alongside right? Is it possible to use one of those out of the box and not make the actor exit until a condition is met (expected document count) rather than after the repeat or duration?